Any CPU Ready Source Inspired by @Chicken-Bones

Akarinnnnn opened the pull request:
Originally posted in [#699](https://github.com/rlabrecque/Steamworks.NET/issues/699#issuecomment-3194215881). Achieved by register a dll resolve hook to ~~default Assembly Load Context~~ Steamworks.NET assembly. 

Dependencies to `static class NativeLibrary` makes this change is not meaningful to any runtimes other than .NET(Core). So these changes are separated into a different folder to keep old version unchanged like .NET Framework 4.0 does.

You will see horrible long commit history because the management of `SteamworksParser` is switched to `git subtree`. Which means I imported `SteamworksParser` files and history into this repository. Since `SteamworksParser` is always bundled with source code generator, in my practice, I finally decided manage them in monorepo way.

Both code generator and generated C# source are tested with v1.62. [v1.63 NuGet package](https://www.nuget.org/packages/Steamworks.NET.AnyCPU/2026.163.0-a) is on public beta.

rlabrecque committed:
Import steamworksparser.py

rlabrecque committed:
Added support for unnamed enums (enum constants)

rlabrecque committed:
Added .gitignore for *.pyc files

rlabrecque committed:
The GameServer APIContext now has a UGC accessor

rlabrecque committed:
Fixed typedef.filename always being blank

rlabrecque committed:
Enable recursive parsing for multiple /**/ comment blocks on one line

rlabrecque committed:
Skip _STEAM_CALLBACK_ Macros

rlabrecque committed:
Added README

rlabrecque committed:
Added MIT license

rlabrecque committed:
ArgAttribute improvements

rlabrecque committed:
Added parsing FunctionAttributes

rlabrecque committed:
Parse STEAM_PRIVATE_API()

rlabrecque committed:
Skip parsing #pragma and #undef

And sorted Skippedlines

rlabrecque committed:
Allow setting ArgAttribute upon creation

rlabrecque committed:
Fixed loading CP-1252 files on linux...

rlabrecque committed:
Added typedefs to comments and made typedefs exist per file.

rlabrecque committed:
Updated GameServerInterfaces

rlabrecque committed:
Updated for Steamworks 1.42

rlabrecque committed:
Updated for Steamworks 1.43

rlabrecque committed:
Update for 1.44, skipping SteamNetworkingSockets for now

rlabrecque committed:
Fix attribute value parsing

rlabrecque committed:
Ignore steamnetworkingsockets.h

rlabrecque committed:
Update for Steamworks SDK 1.48

rlabrecque committed:
Add some support for steamnetworkingtypes.h

rlabrecque committed:
Add support for isteamnetworkingutils.h

rlabrecque committed:
Add support for steamnetworkingsockets.h and steamdatagram_tickets.h

rlabrecque committed:
Add game server versions of isteamnetworkingutils and isteamnetworkingsockets

rlabrecque committed:
Stop parsing isteamcontroller.h

rlabrecque committed:
Update README and LICENSE

rlabrecque committed:
Correctly parse inline function definitions inside interfaces

rlabrecque committed:
Update for Steamworks SDK 1.52

rlabrecque committed:
Update for Steamworks SDK 1.52 - sort-of handle non-callback enum constants inside a struct

rlabrecque committed:
Update for Steamworks SDK 1.55 - skip isteamdualsense.h

TwoPoint-AndyChappell committed:
Add support for the new STEAM_CALLBACK_MEMBER_ARRAY macro

TwoPoint-AndyChappell committed:
Handle the static const int value within GetTicketForWebApiResponse_t
correctly as a constant.  Scan things at scope depth 1 as well as 0 now
to catch it, and also exclude field names in structs that contain an =.
Means the field is no longer added to the struct and is instead added
correctly to constants.

rlabrecque committed:
Add SteamNetworkingMessages as a GameServer Interface

Akarinnnnn committed:
Add Any CPU support, inspired by @Chicken-Bones

Acheived by register a dll resolve hook to default Assembly Load Context.
Changes are seperated into a different folder to keep old version unchanged.

Akarinnnnn committed:
Merge remote-tracking branch 'origin/master' into anycpu-2

Chicken-Bones commented on com.rlabrecque.steamworks.net/Runtime/Steam.cs:
```diff
@@ -111,7 +114,42 @@ public static ESteamAPIInitResult InitEx(out string OutSteamErrMsg)
+
+				string x64LibName = name + "64.dll";
+				// we don't care failed load, let next handler manage it
+				NativeLibrary.TryLoad(x64LibName, out nint lib);
```
This will only work if the 64bit dll is present on the current native search path. It will not call any other `ResolvingUnmanagedDll` callbacks.

You can improve compatibility by:
1. Using `NativeLibrary.TryLoad(x64LibName, assembly, null, out nint lib);`, this is most similar to the default call made when a native library is requested by a managed assembly
2. Using `steam_api64` as the lib name (without the `.dll` suffix). It's not needed for native lib loading, and not passed to `ResolvingUnmanagedDll` in typical use

Akarinnnnn replied:
Good suggestion! I was finding way to maintain search order, but nothing out. Note 1 helped a lot.

Update are pushed. It will be packed into a pre-release later.

Chicken-Bones reviewed (approved):
This is a good implementation, it works perfectly for our case, but I've left a note on how to reduce potential breakage even further. I also tested the nuget package. Thankyou

Akarinnnnn committed:
Imporve Any CPU compatibility

Chicken-Bones commented:
Any update on this one?

Akarinnnnn committed:
Pack steam natives we building against into AnyCPU package

Akarinnnnn commented:
Thanks, I will fix this later.



---Original---
From: ***@***.***&gt;
Date: Fri, Sep 19, 2025 09:27 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [rlabrecque/Steamworks.NET] Less invasive Any CPU implementationinspired by @Chicken-Bones (PR #738)




 
@Chicken-Bones commented on this pull request.
 
 
In com.rlabrecque.steamworks.net/Runtime/Steam.cs:
 &gt; @@ -56,6 +56,9 @@ public static class SteamAPI {  		// Returns true on success  		public static ESteamAPIInitResult InitEx(out string OutSteamErrMsg)  		{ +#if STEAMWORKS_ANYCPU +			BeAnyCPUCompatible();  
You also need to install the hook in RestartAppIfNecessary since docs say to call this before ``SteamAPI.Init`
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

Akarinnnnn committed:
Fix anycpu hook coverage and a typo.

Akarinnnnn commented on com.rlabrecque.steamworks.net/Runtime/Steam.cs:
```diff
@@ -56,6 +56,9 @@ public static class SteamAPI {
 		public static ESteamAPIInitResult InitEx(out string OutSteamErrMsg)
 		{
+#if STEAMWORKS_ANYCPU
+			BeAnyCPUCompatible();
```
Fixed. Also fixed `GameServer` related initialization hooks.

Chicken-Bones commented:
Thanks, please let me know when I can get it via NuGet :)

Akarinnnnn committed:
Changed when to check steam native requester. Fix indentation.

Reasons are explained detailly inside comments. A debug message will also generate if confliction happens.

Akarinnnnn committed:
Explain why not check platform at initialization stage.

Chicken-Bones commented:
#747 

`Packsize.value` is being compiled as 4. 
<img width="993" height="105" alt="image" src="https://github.com/user-attachments/assets/a2276ffe-9133-4187-af49-378838fda271" />

Akarinnnnn commented:
This is a known issue. I made a suggestion to remove this const in previous Any CPU implementation.



---Original---
From: ***@***.***&gt;
Date: Thu, Oct 2, 2025 11:08 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [rlabrecque/Steamworks.NET] Less invasive Any CPU implementationinspired by @Chicken-Bones (PR #738)


Chicken-Bones left a comment (rlabrecque/Steamworks.NET#738)
 
#747
 
Packsize.value is being compiled as 4.
 image.png (view on web)
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

Chicken-Bones commented:
Sinply removing it probably won't work, because it's required to be 4 on osx and linux, despite the default abi for x64 and arm64 on those platforms being 8

Akarinnnnn commented:
> Only easier way would be if you know how to ship different managed assemblies for windows and non-windows in a nuget package...

> it should be possible to make NuGet package structure like

If we want to ship different assemblies for different platform, it's possible, and we **already achieved** this goal, But we have to manually write MSBuild script and NuGet specification file, then execute them by some **long** and **complex** commands, which is not easy to type correct in once. This hard way is probably how they build package `Steamworks.NET`. I don't know how someone use my buggy workflow to build so much package versions, thanks to them.

Return to our subject, when I created this any CPU variant, I decided to simplify packaging workflow into one command: `dotnet pack --version $(Version)`. A simpler and robust way.

To achieve this goal, we must discard manually written `.nuspec`. In result, we can no longer specify where to put our main output(DLL) inside packages. This brings an another bad news: Even if we managed to built all alignment variant, we can't use them for such pack workflow. After some investigation, I found `dotnet pack` can't build all configurations, it uses single one form start to end, typically `Release`.

But `dotnet pack` have another advantages: It will build for all target frameworks and place their output correctly. It will be important when we want to start support more than one target framework.

I have to say, through we can inherit old NuGet workflow from Standalone2.0, this is a chance to embrace revolution. Please let me know which path should we continue. React `❤️` for `dotnet pack --version 2025.162.3`, `👀` for inherit `dotnet build -t:BatchBuild Steamworks.NET.Standard.sln -p:Version=2025.162.3` then windows only `nuget.exe pack Steamworks.NET.nuspec -OutputDirectory bin\`.

Anymore, `Packsize.value` should *never* be a `public const` for any CPU variant(I think other version too). Either hide it or turn it to `static readonly` as #675 said, we should do some act.

Akarinnnnn committed:
Fix most callback system related marshal issues.

Added a conditional marshalling table to decide which alignment to use at runtime, this should fix most callback-retrieve related issues.

Struct parameters in interface methods are still need to fix.

Akarinnnnn committed:
Handle UTF-8 encoded headers better

Akarinnnnn committed:
Try add struct layout analysis, add type annotations

Akarinnnnn committed:
Attempt to make parser smarter

Adding support for nested types and unions, these work are in progress.

Akarinnnnn committed:
Attempt to calculate total size of structs, part 3

Currently I still have to debug unions, they caused nested type stack corrupt I think... I need help

Akarinnnnn committed:
Attempt to calculate total size of structs, part 4

Nested type parsing and some special structs are ok I think. Next i will work on calculation.

Akarinnnnn committed:
Attempt to calculate total size of structs, attempt 5

Is it work?

Akarinnnnn committed:
[Feature] Implement struct field offset analysis.

This feature can be used to generate helper structs that need to generate large and small pack variant. This problem is caused by default pack-size make the struct have different layout on different platforms.

Generate these struct with both variant is very important to building a unified Any CPU Steamworks.NET assembly. After this commit

Akarinnnnn committed:
Preparing use subtree instead of submodule

Akarinnnnn committed:
Add 'CodeGen/SteamworksParser/' from commit '665fb115a9182883457a6bb565d1c3b7c07a8c42'

git-subtree-dir: CodeGen/SteamworksParser
git-subtree-mainline: cffcf0484ed3d73872d6bd577be081dd36eed5e1
git-subtree-split: 665fb115a9182883457a6bb565d1c3b7c07a8c42

Akarinnnnn committed:
Make struct layout analysis work

Akarinnnnn committed:
Add support to hack marshalling on demand

This commit contains only codegen

Akarinnnnn committed:
Hide const packsize, expose it by static readonly field instead

Akarinnnnn committed:
Allow generator to ignore structs, clearify ambiguous type of `Field.arraysize` by Python type annotation

Akarinnnnn committed:
Source geneator now generate both native function variant for packsize aware structs.

Corresponding wrapper method is pending.

Akarinnnnn committed:
Source generator now generate both variant of packsize aware struct.

Akarinnnnn committed:
Generated code not work, need fix.

There is a conflict constant `k_cchMaxString` caused by  `SteamNetworkingIdentity` and `SteamNetworkingIPAddr` defines same named field, in two different anonymous enum

Akarinnnnn committed:
Fixed constant value confliction by ignore a problemtic file.

Types inside struct and `SteamParamStringArray_t` may get fixed today(in UTC+8)

Akarinnnnn committed:
Fix steam headers parser for ignoring `SteamParamStringArray_t::m_ppStrings`

Akarinnnnn committed:
Fix `Callback<T>` doesn't marshal Any CPU correctly.

Interface parameter usage of struct still need work.

Akarinnnnn committed:
Make SrcGen can generate Any CPU interface wrapper.

Native methods are still in progress. First is LP/SP native function pair `ISteamParties_GetBeaconLocationData` has been duplicated twice, it seems same function generated twice.

Secondly large pack `ISteamParties_GetAvailableBeaconLocations`'s second arg, `p_locationList_lp`, 's associated type is not properly generated. Array brackets is generated before `_LargePack`. I' ll try to find out why.

Akarinnnnn committed:
Fixed pack-size aware typed parameter array generation.

Akarinnnnn committed:
Regenerate native imports with array fix.

Akarinnnnn committed:
Merge remote-tracking branch 'origin/master' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Fix gameserver interfaces generation

Akarinnnnn committed:
All Steam interface wrappers seems work.

Akarinnnnn committed:
Regenerate steam interface wrappers.

Akarinnnnn committed:
Fix array parameter usage of packsize aware structs.

Regenerated files are also included.

ryan-linehan committed:
add a github workflow to pack and push nuget package

ryan-linehan committed:
update workflow to take inputs for version and publishing

ryan-linehan committed:
update action name

ryan-linehan committed:
add tag versioning for publishing nuget package + add approver step for 'prod' environment

ryan-linehan committed:
default version to 0.0.0 so build can still run on push

ryan-linehan committed:
remove branch trigger on push to + utilize proper variable syntax for nuget user

Akarinnnnn committed:
Add workflow for packing and publishing steamworks Steamworks.NET.AnyCPU (#2)

Adds a simple workflow for packing the Steamworks.NET.AnyCPU nuget
package and publishing it to nuget.org. Currently only run on manual run
of the workflow.

The workflow uses nuget.org's newer [trusted publishing
](https://learn.microsoft.com/en-us/nuget/nuget-org/trusted-publishing)
so there is a step for the owner of the nuget package on nuget.org to do
as well as the action variable of NUGET_USER. Its not a secret but this
allows another to use the action without it hard coded. Ideally, we get
to be able to publish a beta branch through the workflow to make sure it
works.

Akarinnnnn committed:
Merge remote-tracking branch 'upstream/master' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Merge upstream changes.

Merge changes from upstream to keep us up to date.

It's here because #3 merged into a legacy branch master.

ryan-linehan committed:
include STEAMWORKS_WIN constant

ryan-linehan committed:
add a fix to dynamically load assembly on windows

Akarinnnnn committed:
Revert adding of macro `STEAMWORKS_WIN`

To avoid breaking internal symbol usages, `STEAMWORKS_WIN` is dropped versus `STEAMWORKS_LIN_OSX`.

Akarinnnnn committed:
Changed the time to hook dll loading.

Now we can do our hook in any AssemblyLoacContext s. DLL loading issues in godot(editor?)  caused by multiple A.L.C.s should be fixed.

Akarinnnnn committed:
Move loader hook fix to source generator's template.

Akarinnnnn committed:
Move loader hook fix to source generator's template.

Akarinnnnn committed:
Change the place of DLL resolve hook (#6)

Fixed #5 by move inject DLL loader hook to `NativeLibrary` provided interception point. This allows we hook loading process in any ALC, which is especially important for Godot.

Akarinnnnn committed:
Fix anycpu large pack native function was leak.

Akarinnnnn committed:
Fix non-anycpu variant uses anycpu wrapper

It will cause complication failure.

Akarinnnnn committed:
Regenreate to fix AnyCPU wrappers leak.

ryan-linehan committed:
update sendsignal to be pack aware with manual edits

ryan-linehan committed:
update structs.py to include SteamNetConnectionInfo_t as a sequential struct

ryan-linehan committed:
update send signal template to be pack size aware

ryan-linehan committed:
update nativemethods.txt to have pack aware dll import

Akarinnnnn committed:
Fix packsize aware struct nested in pack aware struct.

But pack aware struct nested in common struct are not in consideration. Write it down here for one day we met some bug related.

Akarinnnnn committed:
Create bug report template

Akarinnnnn committed:
Create bug report template (#11)

Akarinnnnn committed:
SendSignal fixes without changing generated files (#9)

A workaround of incorrect codegen of `SteamNetConnectionInfo_t` .

Akarinnnnn committed:
Show pack and size of structs by C++

Akarinnnnn committed:
Pull from latest contributions

Akarinnnnn committed:
Merge branch 'sendsignal_gen_fixes' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Merge branch 'anycpu-packable-fix-align-by-conditionalmarshal' of https://github.com/Akarinnnnn/Steamworks.NET.AnyCPU into anycpu-packable-fix-align-by-conditionalmarshal

ropemonkey committed:
Fix invalid escape sequence warnings by converting regex patterns to raw strings

Akarinnnnn committed:
Merge branch 'anycpu-2' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Commit parser to be merged to packable branch

Akarinnnnn committed:
Merge branch 'sendsignal_gen_fixes' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Try to analyze structs like CGameID

Puropse is to exclude fake top level structs like CGameID::GameID_t. Maybe just add a table is enough?

Akarinnnnn committed:
Add an comparer to official generated data.

Akarinnnnn committed:
Revert class parsing and fixed array related field handling

Akarinnnnn committed:
Simplify marshal fix review (#12)

Simplify marshal fix #1 's review page, hide these `git subtree`
imported commits from by merge them into base branch `anycpu-2`.

Akarinnnnn committed:
Merge branch 'anycpu-2' into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Add a advanced script to test parser's pack-awareness

Akarinnnnn committed:
Regenerate after struct field scan algorithm update

Akarinnnnn committed:
Merge branch 'anycpu-packable-fix-align-by-conditionalmarshal' of https://github.com/Akarinnnnn/Steamworks.NET.AnyCPU into anycpu-packable-fix-align-by-conditionalmarshal

Akarinnnnn committed:
Changed DllImportResolver hack to allow searching .so etc. in assembly directory

Akarinnnnn committed:
Fixed old win64 steam native resolve logic

Akarinnnnn committed:
Fix large pack structs weren't being used

Akarinnnnn committed:
Merge upstream SteamworksParser changes

Akarinnnnn committed:
Minor misc changes

Update pre-baked generator resource.

Make `SteamMarshallerTable` public.

Added `.editorconfig` to enforce style.

Akarinnnnn committed:
Add missing "lib" prefix to native resolve logic

Move Any CPU native resolve logic to another file.

Minor changes to .editorconfig

Akarinnnnn committed:
Fix Godot loading compatibility

Akarinnnnn committed:
Fix pack size related marshal issues.

Our goal is to fix most callback system related marshal issues.

Added a conditional marshalling table to decide which alignment to use
at runtime, this should fix most callback-retrieve related issues..

Parameter marshal is done, through it looks ugly. It will be
fixed in next upcoming PR.

We have a mechanism to find out packsize aware structs now.

Akarinnnnn committed:
Bind to Steamworks SDK v1.63

Minor changes from upstream

Akarinnnnn committed:
Sync upstream with new target: Steamworks SDK v1.63 (#14)

Sync upstream with new target: Steamworks SDK v1.63

Some minor changes also included.

Akarinnnnn commented:
Both code generator and generated C# source are tested with v1.62. [v1.63 NuGet package](https://www.nuget.org/packages/Steamworks.NET.AnyCPU/2026.163.0-a) is on public beta.

Akarinnnnn committed:
ver 2025.163.1-a

Add android and linux arm64 binaries

Changed <PackageProjectUrl> and <RepositoryUrl> to fork

Akarinnnnn committed:
NuGet package and metadata changes (#15)

Add android and linux arm64 binaries

Changed <PackageProjectUrl> and <RepositoryUrl> to here

Akarinnnnn committed:
Bump Steamworks SDK to v1.64

Akarinnnnn committed:
Upgrade Steamworks SDK to v1.64 (#17)

Upgrade Steamworks SDK to v1.64

Akarinnnnn committed:
Restructure AnyCPU project, rebind, and contributor guidelines for AnyCPU

Akarinnnnn committed:
[v2026.142.1] Restructure AnyCPU project, expose `ISteamCallbackIdentity` (#18)

Restructure project files for clearer solution view, expose callback
identity interface for extensibility, add contributor guideline for
future.

Akarinnnnn committed:
Remove old file

Akarinnnnn committed:
Fix load issues on Godot 4.7.1 Win64

Akarinnnnn committed:
[v2026.164.3] Fix Godot Win64 load issues (#20)

Fixed #19 Steam DLL Resolution failed on Godot Windows
